beacon-chain/sync/pending_attestations_queue.go

```diff
@@ -18,161 +17,144 @@ import (
+				log.WithError(err).Debug("Could not retrieve attestation prestate")
+				continue
+			}
+
+			valid, err := s.validateUnaggregatedAttWithState(ctx, att.Aggregate, preState)
+			if err != nil {
+				log.WithError(err).Debug("Pending unaggregated attestation failed validation")
+				continue
+			}
+			if valid == pubsub.ValidationAccept {
+				if err := s.cfg.AttPool.SaveUnaggregatedAttestation(att.Aggregate); err != nil {
+					log.WithError(err).Debug("Could not save unaggregated attestation")
+					continue
+				}
+				s.setSeenCommitteeIndicesSlot(att.Aggregate.Data.Slot, att.Aggregate.Data.CommitteeIndex, att.Aggregate.AggregationBits)
+
+				valCount, err := helpers.ActiveValidatorCount(ctx, preState, slots.ToEpoch(att.Aggregate.Data.Slot))
+				if err != nil {
+					log.WithError(err).Debug("Could not retrieve active validator count")
+					continue
+				}
+				// Broadcasting the signed attestation again once a node is able to process it.
+				if err := s.cfg.P2P.BroadcastAttestation(ctx, helpers.ComputeSubnetForAttestation(valCount, signedAtt.Message.Aggregate), signedAtt.Message.Aggregate); err != nil {
+					log.WithError(err).Debug("Could not broadcast")
+				}
+			}
 		}
 	}
-	return s.sendBatchRootRequest(ctx, pendingRoots, randGen)
+
+	log.WithFields(logrus.Fields{
+		"blockRoot":        hex.EncodeToString(bytesutil.Trunc(bRoot[:])),
+		"pendingAttsCount": len(attestations),
+	}).Debug("Verified and saved pending attestations to pool")
+
+	// Delete the missing block root key from pending attestation queue so a node will not request for the block again.
+	s.pendingAttsLock.Lock()
+	delete(s.blkRootToPendingAtts, bRoot)
+	s.pendingAttsLock.Unlock()
 }
 
 // This defines how pending attestations is saved in the map. The key is the
 // root of the missing block. The value is the list of pending attestations
-// that voted for that block root.
-func (s *Service) savePendingAtt(att *ethpb.SignedAggregateAttestationAndProof) {
+// that voted for that block root. If the key is not in blkRootToPendingAtts
+// then a request for the block is sent.
+func (s *Service) savePendingAtt(att *ethpb.SignedAggregateAttestationAndProof) error {
 	root := bytesutil.ToBytes32(att.Message.Aggregate.Data.BeaconBlockRoot)
 
 	s.pendingAttsLock.Lock()
-	defer s.pendingAttsLock.Unlock()
 	_, ok := s.blkRootToPendingAtts[root]
 	if !ok {
+		randGen := rand.NewGenerator()
+		// Is the first time we save an attestation for this block,
+		// we request it.
 		s.blkRootToPendingAtts[root] = []*ethpb.SignedAggregateAttestationAndProof{att}
-		return
+		s.pendingAttsLock.Unlock()
+
+		log.WithFields(logrus.Fields{
+			"currentSlot": s.cfg.Chain.CurrentSlot(),
+			"attSlot":     att.Message.Aggregate.Data.Slot,
+			"blockRoot":   hex.EncodeToString(bytesutil.Trunc(root[:])),
+		}).Debug("Requesting block for pending attestation")
+
+		return s.sendBatchRootRequest(context.Background(), [][32]byte{root}, randGen)
 	}
 
+	defer s.pendingAttsLock.Unlock()
```

Comment:
This will be a deadlock if we exit early above when sending the batch request.

Reply:
don't I unlock explicitly in that case?